firedrake/mesh.py

```diff
@@ -2282,6 +2310,59 @@ def mark_entities(self, f, label_name, label_value):
         """
         self.topology.mark_entities(f.topological, label_name, label_value)
 
+    @utils.cached_property  # TODO: Recalculate if mesh moves
+    def input_ordering(self):
+        """
+        Return the input ordering of the mesh vertices as a
+        :func:`~.VertexOnlyMesh` whilst preserving other information, such as
+        the global indices and parent mesh cell information.
+
+        Only available for vertex-only meshes.
+
+        Notes
+        -----
+        If ``redundant=True`` at mesh creation, all the vertices will
+        be returned on rank 0.
+
+        Any points that were not found in the original mesh when it was created
+        will still be present here in their originally supplied order.
+        """
+        if not isinstance(self.topology, VertexOnlyMeshTopology):
+            raise AttributeError("Input ordering is only defined for vertex-only meshes.")
+        return self._input_ordering
+
+    @utils.cached_property  # TODO: Recalculate if mesh moves
+    def input_ordering_sf(self):
+        """
+        Return a PETSc SF which has :func:`~.VertexOnlyMesh` input ordering
+        vertices as roots and this mesh's vertices as leaves.
```

Comment:
When we (re)distribute a `plex` to get `distplex` using `DMPlexDistribute`, it returns an SF called  `distsf`, whose roots and leaves are points in `plex` and those in `distplex`. I think a `swarm` and what you call `input_ordering` are very similar to `plex` and `distplex`, and `input_ordering_sf` compares to `distsf` (though `distsf` is from point to point, while `input_ordering_sf` is from firedrake_number to firedrake_number ?) So I think it is more natural to have something like:
```
class VertexOnlyMeshTopology():
    ...
    self._distribution_parent = ...  # `None` for what you currently call `input_ordering` and
                                                  # `input_ordering` for the vom
    self._distribution_sf = None # or what you call `input_ordering_sf`
```
If we ever consider distributing swam points more than once, we could have:
```
    def distribution_ancesters():
        if self._distribution_parent:
            return [self] + self._distribution_parent.distribution_ancesters()
        else:
            return [self]

    def distribution_sf():
        if self._distribution_parent:
            return self._distribution_parent.distribution_sf().compose(self._distribution_sf)
        else:
            return self._distribution_sf
```

Reply:
I'm not sure if these are directly comparable. Even in serial, the input ordering VOM includes points which are not found in the parent mesh (assuming you've turned off the error for this). Also, given this is very much a user facing part of the API, I think `input_ordering` is perhaps more expressive than `distribution_parent`, particularly given the point I just made?

I like your idea about being able to redistribute these in the future though - when we do that we can make `input_ordering` just return the first distribution ancestor. I question about your suggestion however, it looks like here
```python
    def distribution_ancesters():
        if self._distribution_parent:
            return [self] + self._distribution_parent.distribution_ancesters()
```
you're adding VOMs together - I don't think that's possible?

Reply:
That is a list addition, so that is to return a list of ancestors, but anyway, I see. Let's leave it as is now.

Reply:
Oh of course, I missed the `[]`!